src/Stack/Setup.hs

```diff
@@ -1821,6 +1827,69 @@ downloadStackExe platforms0 archiveInfo destDir testExe = do
             let base = FP.dropExtension (FP.takeBaseName (T.unpack url)) FP.</> "stack"
              in if isWindows then base FP.<.> "exe" else base
 
+-- | Ensure that the Stack executable download is in the same location
+-- as the currently running executable. See:
+-- https://github.com/commercialhaskell/stack/issues/3232
+performPathChecking
+    :: (MonadIO m, MonadLogger m, MonadCatch m, MonadReader env m, HasConfig env)
+    => Path Abs File -- ^ location of the newly downloaded file
+    -> m ()
+performPathChecking newFile = do
+  executablePath <- liftIO getExecutablePath
+  executablePath' <- parseAbsFile executablePath
+  unless (toFilePath newFile == executablePath) $ do
+    $logInfo $ T.pack $ "Also copying stack executable to " ++ executablePath
+    tmpFile <- parseAbsFile $ executablePath ++ ".tmp"
+    eres <- tryIO $ do
+      liftIO $ copyFile newFile tmpFile
+#if !WINDOWS
+      liftIO $ setFileMode (toFilePath tmpFile) 0o755
+#endif
+      liftIO $ renameFile tmpFile executablePath'
+      $logInfo "Stack executable copied successfully!"
+    case eres of
+      Right () -> return ()
+      Left e
+        | isPermissionError e -> do
+            $logWarn $ T.pack $ "Permission error when trying to copy: " ++ show e
+            $logWarn "Should I try to perform the file copy using sudo? This may fail"
```

Comment:
It might be good to show the users exactly which commands will be run (that way they have the option to ctrl-C and do it themselves if they don't trust stack to run commands as root itself)